Do not inappropriately register bind when the value is a closure

vlakoff opened the pull request:
Prerequisite for #5138.

The main fix is commit https://github.com/vlakoff/CodeIgniter4/commit/223ca28400a95594490052b032e1a921237138fb.

kenjis commented:
Unit testing tests the object behavior, not in the code, not the specific method.

kenjis commented:
According to line coverage, there are two lines 682,683 in `whereHaving()` that are not covered. 
![Screenshot 2021-10-30 11 40 25](https://user-images.githubusercontent.com/87955/139517865-2d823d05-90f1-405e-a8d7-1a23fc5af54b.png)

kenjis commented:
The branch coverage for `whereHaving()` is 84.38%.
Not bad.

```
$ ./phpunit --coverage-html=build/coverage --path-coverage tests/system/Database/Builder/ -d memory_limit=1024m
```

![Screenshot 2021-10-30 13 15 55](https://user-images.githubusercontent.com/87955/139519864-dfc0b836-946a-4799-9a9f-2e4f35d6b818.png)

kenjis commented:
@vlakoff Adding this test gets 100% line coverage.

```php
    public function testWhereAssociateArrayKeyHasEqualValueIsNull()
    {
        $builder = $this->db->table('user');

        $where = [
            'id <'   => 100,
            'col1 =' => null,
        ];
        $builder->where($where);

        $expectedSQL = 'SELECT * FROM "user" WHERE "id" < 100 AND "col1" IS NULL';
        $this->assertSame($expectedSQL, str_replace("\n", ' ', $builder->getCompiledSelect()));
    }
```

kenjis commented:
This PR is very good.
The commit messages are great, with clear explanations!

vlakoff commented:
I have added this unit test. I made some changes starting from the code you provided:

* put only the "where" condition we are specifically testing
* added "$expectedBinds", for the sake of completeness and to show a bind isn't registered

kenjis commented:
I tried to think of a use case to test vlakoff@4f702d9, but couldn't come up with one.

kenjis commented:
Okay, line coverage 100%, Branch coverage 93.75%.

![Screenshot 2021-10-30 18 49 39](https://user-images.githubusercontent.com/87955/139528262-d622e655-c626-4e35-90f8-32c195208c89.png)

kenjis reviewed (approved):
I looked at all the commits, and it looks good to me.

paulbalandan commented on system/Database/BaseBuilder.php:
```diff
@@ -659,33 +659,34 @@ protected function whereHaving(string $qbKey, $key, $value = null, string $type
 
                     if (substr($k, -1 * strlen($op)) === $op) {
-                        $k = rtrim(strrev(preg_replace(strrev('/' . $op . '/'), strrev(''), strrev($k), 1)));
+                        $k  = rtrim(substr($k, 0, -1 * strlen($op)));
```
```suggestion
                        $k  = rtrim(substr($k, 0, -strlen($op)));
```

paulbalandan commented on system/Database/BaseBuilder.php:
```diff
@@ -659,33 +659,34 @@ protected function whereHaving(string $qbKey, $key, $value = null, string $type
             } elseif (preg_match('/\s*(!?=|<>|IS(?:\s+NOT)?)\s*$/i', $k, $match, PREG_OFFSET_CAPTURE)) {
-                $k = substr($k, 0, $match[0][1]) . ($match[1][0] === '=' ? ' IS NULL' : ' IS NOT NULL');
+                $k  = substr($k, 0, $match[0][1]);
+                $op = ($match[1][0] === '=' ? ' IS NULL' : ' IS NOT NULL');
```
```suggestion
                $op = $match[1][0] === '=' ? ' IS NULL' : ' IS NOT NULL';
```

vlakoff commented on system/Database/BaseBuilder.php:
```diff
@@ -659,7 +659,7 @@ protected function whereHaving(string $qbKey, $key, $value = null, string $type
                     $op = trim(current($op));
 
                     if (substr($k, -1 * strlen($op)) === $op) {
```
The same change `-strlen` should be applied here as well.

vlakoff commented:
Force-pushed, integrating the small code style changes from reviews.

kenjis commented:
@vlakoff Thank you!
